File: agent/src/autosecrets_agent/envelope.py
"""Versioned Agent envelope protocol v1.

Mirror of the Go reference implementation in core/internal/envelope. Both
implementations must produce identical canonical JSON bytes (sorted keys, no
whitespace, Go-compatible string escaping) so signatures and manifests verify
across languages. See api/agent-envelope/envelope-v1.md.
"""

from __future__ import annotations

import base64
import hashlib
import io
import json
from dataclasses import asdict, dataclass
from datetime import UTC, datetime
from typing import Any

from age.file import Decryptor, Encryptor
from age.keys.agekey import AgePrivateKey, AgePublicKey
from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey,
    Ed25519PublicKey,
)

PROTOCOL_NAME = "autosecrets-envelope"
VERSION = 1
SUITE_AGE_X25519 = "age-x25519"

_RFC3339_UTC = "%Y-%m-%dT%H:%M:%SZ"


class EnvelopeError(Exception):
    """Base class for all envelope failures. Callers must treat these as
    fail-closed outcomes, never as a signal to continue."""


class UnsupportedProtocolError(EnvelopeError):
    pass


class UnsupportedVersionError(EnvelopeError):
    pass


class UnsupportedSuiteError(EnvelopeError):
    pass


class ExpiredError(EnvelopeError):
    pass


class BadSignatureError(EnvelopeError):
    pass


class CiphertextError(EnvelopeError):
    pass


class BadManifestError(EnvelopeError):
    pass


class CanonicalError(EnvelopeError):
    pass


def canonical_json(obj: Any) -> str:
    """Serialize like Go's encoding/json: sorted keys, no whitespace, raw
    non-ASCII bytes, and <, >, &, U+2028, U+2029 escaped exactly as Go does."""
    text = json.dumps(obj, sort_keys=True, separators=(",", ":"), ensure_ascii=False)
    return (
        text.replace("<", "\\u003c")
        .replace(">", "\\u003e")
        .replace("&", "\\u0026")
        .replace("\u2028", "\\u2028")
        .replace("\u2029", "\\u2029")
    )


def canonical_manifest(files: list[dict[str, str]]) -> bytes:
    """Serialize FileSpecs in the canonical manifest form."""
    try:
        entries = [dict(f) for f in sorted(files, key=lambda f: f["path"])]
        for entry in entries:
            for key in ("gid", "mode", "path", "sha256", "uid"):
                if key not in entry:
                    raise KeyError(key)
    except (KeyError, TypeError) as exc:
        raise CanonicalError(f"invalid file spec: {exc}") from exc
    manifest = {"files": entries, "protocol": "autosecrets-manifest", "version": "1"}
    return canonical_json(manifest).encode()


@dataclass
class Envelope:
    protocol: str
    version: int
    node_id: str
    revision_id: str
    created_at: str
    expires_at: str
    manifest_sha256: str
    suite: str
    ciphertext: str
    signing_key_id: str
    signature: str

    @classmethod
    def from_json(cls, data: str | bytes) -> Envelope:
        obj = json.loads(data)
        return cls(**obj)

    def to_json(self) -> str:
        return json.dumps(asdict(self), sort_keys=True, separators=(",", ":"))

    def signature_payload(self) -> bytes:
        fields = {
            "ciphertext": self.ciphertext,
            "created_at": self.created_at,
            "expires_at": self.expires_at,
            "manifest_sha256": self.manifest_sha256,
            "node_id": self.node_id,
            "protocol": self.protocol,
            "revision_id": self.revision_id,
            "signing_key_id": self.signing_key_id,
            "suite": self.suite,
            "version": str(self.version),
        }
        return canonical_json(fields).encode()

    def _check_meta(self) -> None:
        if self.protocol != PROTOCOL_NAME:
            raise UnsupportedProtocolError(f"unsupported protocol {self.protocol!r}")
        if self.version != VERSION:
            raise UnsupportedVersionError(f"unsupported version {self.version!r}")
        if self.suite != SUITE_AGE_X25519:
            raise UnsupportedSuiteError(f"unsupported suite {self.suite!r}")

    def _check_expiry(self, now: datetime) -> None:
        if not self.expires_at:
            return
        try:
            expiry = datetime.strptime(self.expires_at, _RFC3339_UTC).replace(tzinfo=UTC)
        except ValueError as exc:
            raise EnvelopeError(f"invalid expires_at {self.expires_at!r}") from exc
        if now > expiry:
            raise ExpiredError(f"envelope expired at {self.expires_at}")

    def verify(self, verify_pub: bytes, now: datetime) -> None:
        self._check_meta()
        self._check_expiry(now)
        try:
            signature = base64.b64decode(self.signature, validate=True)
        except (ValueError, TypeError) as exc:
            raise BadSignatureError("signature is not valid base64") from exc
        try:
            public_key = Ed25519PublicKey.from_public_bytes(verify_pub)
            public_key.verify(signature, self.signature_payload())
        except (InvalidSignature, ValueError) as exc:
            raise BadSignatureError("signature does not verify") from exc

    def open(self, identity: AgePrivateKey, verify_pub: bytes, now: datetime) -> bytes:
        self.verify(verify_pub, now)
        try:
            ciphertext = base64.b64decode(self.ciphertext, validate=True)
        except (ValueError, TypeError) as exc:
            raise CiphertextError("ciphertext is not valid base64") from exc
        try:
            decryptor = Decryptor([identity], io.BytesIO(ciphertext))
            return decryptor.read()
        except Exception as exc:
            raise CiphertextError("ciphertext could not be decrypted") from exc

    def verify_manifest(self, manifest: bytes) -> None:
        digest = hashlib.sha256(manifest).hexdigest()
        if digest != self.manifest_sha256:
            raise BadManifestError("manifest hash mismatch")


def _format_utc(value: datetime | None) -> str:
    if value is None:
        return ""
    return value.astimezone(UTC).strftime(_RFC3339_UTC)


def create(
    *,
    plaintext: bytes,
    manifest: bytes,
    recipient_pub: AgePublicKey,
    signer: Ed25519PrivateKey,
    node_id: str,
    revision_id: str,
    created_at: datetime | None,
    expires_at: datetime | None,
    signing_key_id: str,
) -> Envelope:
    if not manifest:
        raise EnvelopeError("manifest required")
    if created_at is None:
        created_at = datetime.now(UTC)

    buffer = io.BytesIO()
    with Encryptor([recipient_pub], buffer) as encryptor:
        encryptor.write(plaintext)

    manifest_sha256 = hashlib.sha256(manifest).hexdigest()
    envelope = Envelope(
        protocol=PROTOCOL_NAME,
        version=VERSION,
        node_id=node_id,
        revision_id=revision_id,
        created_at=_format_utc(created_at),
        expires_at=_format_utc(expires_at),
        manifest_sha256=manifest_sha256,
        suite=SUITE_AGE_X25519,
        ciphertext=base64.b64encode(buffer.getvalue()).decode(),
        signing_key_id=signing_key_id,
        signature="",
    )
    envelope.signature = base64.b64encode(signer.sign(envelope.signature_payload())).decode()
    return envelope

File: core/internal/envelope/envelope.go
// Package envelope implements the versioned Agent envelope protocol: a JSON
// object carrying a Secret bundle payload encrypted to one Managed Node's age
// X25519 key and signed by a Core Ed25519 key. Go (Core) and Python (Agent)
// must produce identical canonical signature payloads; see
// api/agent-envelope/envelope-v1.md and the checked-in test vectors.
package envelope

import (
	"bytes"
	"crypto/ed25519"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"sort"
	"strconv"
	"strings"
	"time"

	"filippo.io/age"
)

const (
	ProtocolName   = "autosecrets-envelope"
	Version        = 1
	SuiteAgeX25519 = "age-x25519"
)

var (
	ErrUnsupportedProtocol = errors.New("envelope: unsupported protocol")
	ErrUnsupportedVersion  = errors.New("envelope: unsupported version")
	ErrUnsupportedSuite    = errors.New("envelope: unsupported encryption suite")
	ErrExpired             = errors.New("envelope: expired")
	ErrBadSignature        = errors.New("envelope: bad signature")
	ErrCiphertext          = errors.New("envelope: ciphertext could not be decrypted")
	ErrBadManifest         = errors.New("envelope: manifest hash mismatch")
)

// Envelope is the wire format defined in api/agent-envelope/envelope-v1.md.
type Envelope struct {
	Protocol       string `json:"protocol"`
	Version        int    `json:"version"`
	NodeID         string `json:"node_id"`
	RevisionID     string `json:"revision_id"`
	CreatedAt      string `json:"created_at"`
	ExpiresAt      string `json:"expires_at"`
	ManifestSHA256 string `json:"manifest_sha256"`
	Suite          string `json:"suite"`
	Ciphertext     string `json:"ciphertext"`
	SigningKeyID   string `json:"signing_key_id"`
	Signature      string `json:"signature"`
}

// Options configures New. CreatedAt may be zero to use the current time;
// ExpiresAt zero means the envelope never expires.
type Options struct {
	NodeID       string
	RevisionID   string
	CreatedAt    time.Time
	ExpiresAt    time.Time
	Manifest     []byte
	Plaintext    []byte
	Recipient    *age.X25519Recipient
	Signer       ed25519.PrivateKey
	SigningKeyID string
}

// FileSpec describes one file a Bundle Revision materializes. All values are
// strings so Go and Python serialize them identically.
type FileSpec struct {
	Path   string
	Mode   string
	UID    string
	GID    string
	SHA256 string
}

// signaturePayload returns the canonical bytes covered by the Ed25519
// signature: every envelope field except signature, alphabetically sorted,
// no whitespace, all values strings.
func (e *Envelope) signaturePayload() []byte {
	fields := map[string]string{
		"ciphertext":      e.Ciphertext,
		"created_at":      e.CreatedAt,
		"expires_at":      e.ExpiresAt,
		"manifest_sha256": e.ManifestSHA256,
		"node_id":         e.NodeID,
		"protocol":        e.Protocol,
		"revision_id":     e.RevisionID,
		"signing_key_id":  e.SigningKeyID,
		"suite":           e.Suite,
		"version":         strconv.Itoa(e.Version),
	}
	// encoding/json marshals map keys in sorted order, matching Python's
	// json.dumps(sort_keys=True, separators=(",", ":")).
	data, _ := json.Marshal(fields)
	return data
}

// New encrypts Plaintext to Recipient and signs the resulting envelope with
// Signer. It never logs or persists plaintext.
func New(o Options) (*Envelope, error) {
	if len(o.Manifest) == 0 {
		return nil, errors.New("envelope: manifest required")
	}
	if o.Recipient == nil {
		return nil, errors.New("envelope: recipient required")
	}
	if len(o.Signer) != ed25519.PrivateKeySize {
		return nil, errors.New("envelope: signer required")
	}

	var ciphertext bytes.Buffer
	writer, err := age.Encrypt(&ciphertext, o.Recipient)
	if err != nil {
		return nil, err
	}
	if _, err := writer.Write(o.Plaintext); err != nil {
		return nil, err
	}
	if err := writer.Close(); err != nil {
		return nil, err
	}

	created := o.CreatedAt
	if created.IsZero() {
		created = time.Now().UTC()
	}
	expires := ""
	if !o.ExpiresAt.IsZero() {
		expires = o.ExpiresAt.UTC().Format(time.RFC3339)
	}
	manifestSum := sha256.Sum256(o.Manifest)

	env := &Envelope{
		Protocol:       ProtocolName,
		Version:        Version,
		NodeID:         o.NodeID,
		RevisionID:     o.RevisionID,
		CreatedAt:      created.UTC().Format(time.RFC3339),
		ExpiresAt:      expires,
		ManifestSHA256: hex.EncodeToString(manifestSum[:]),
		Suite:          SuiteAgeX25519,
		Ciphertext:     base64.StdEncoding.EncodeToString(ciphertext.Bytes()),
		SigningKeyID:   o.SigningKeyID,
	}
	signature := ed25519.Sign(o.Signer, env.signaturePayload())
	env.Signature = base64.StdEncoding.EncodeToString(signature)
	return env, nil
}

func (e *Envelope) checkMeta() error {
	if e.Protocol != ProtocolName {
		return ErrUnsupportedProtocol
	}
	if e.Version != Version {
		return ErrUnsupportedVersion
	}
	if e.Suite != SuiteAgeX25519 {
		return ErrUnsupportedSuite
	}
	return nil
}

func (e *Envelope) checkExpiry(now time.Time) error {
	if e.ExpiresAt == "" {
		return nil
	}
	// Both implementations accept exactly "2006-01-02T15:04:05Z" (UTC, Z
	// suffix only) so a Core-produced envelope always parses on the Agent.
	expiry, err := time.Parse("2006-01-02T15:04:05Z", e.ExpiresAt)
	if err != nil {
		return fmt.Errorf("envelope: invalid expires_at: %w", err)
	}
	if now.After(expiry) {
		return ErrExpired
	}
	return nil
}

// Verify checks protocol, version, suite, expiry, and the Ed25519 signature
// without decrypting the payload.
func (e *Envelope) Verify(verifyKey ed25519.PublicKey, now time.Time) error {
	if err := e.checkMeta(); err != nil {
		return err
	}
	if err := e.checkExpiry(now); err != nil {
		return err
	}
	if len(verifyKey) != ed25519.PublicKeySize {
		return ErrBadSignature
	}
	signature, err := base64.StdEncoding.DecodeString(e.Signature)
	if err != nil {
		return ErrBadSignature
	}
	if !ed25519.Verify(verifyKey, e.signaturePayload(), signature) {
		return ErrBadSignature
	}
	return nil
}

// Open verifies the envelope and decrypts the payload with identity.
func (e *Envelope) Open(identity age.Identity, verifyKey ed25519.PublicKey, now time.Time) ([]byte, error) {
	if err := e.Verify(verifyKey, now); err != nil {
		return nil, err
	}
	ciphertext, err := base64.StdEncoding.DecodeString(e.Ciphertext)
	if err != nil {
		return nil, ErrCiphertext
	}
	reader, err := age.Decrypt(strings.NewReader(string(ciphertext)), identity)
	if err != nil {
		return nil, ErrCiphertext
	}
	plaintext, err := io.ReadAll(reader)
	if err != nil {
		return nil, ErrCiphertext
	}
	return plaintext, nil
}

// VerifyManifest checks that manifest hashes to the envelope's
// manifest_sha256.
func (e *Envelope) VerifyManifest(manifest []byte) error {
	sum := sha256.Sum256(manifest)
	if hex.EncodeToString(sum[:]) != e.ManifestSHA256 {
		return ErrBadManifest
	}
	return nil
}

// CanonicalManifest serializes FileSpecs in the canonical manifest form:
// sorted top-level keys, files sorted by path, every value a string.
func CanonicalManifest(files []FileSpec) ([]byte, error) {
	sorted := append([]FileSpec(nil), files...)
	sort.Slice(sorted, func(i, j int) bool { return sorted[i].Path < sorted[j].Path })
	entries := make([]map[string]string, 0, len(sorted))
	for _, f := range sorted {
		entries = append(entries, map[string]string{
			"gid":    f.GID,
			"mode":   f.Mode,
			"path":   f.Path,
			"sha256": f.SHA256,
			"uid":    f.UID,
		})
	}
	manifest := map[string]any{
		"files":    entries,
		"protocol": "autosecrets-manifest",
		"version":  "1",
	}
	return json.Marshal(manifest)
}
